sql: fix stats jobs when they are retried automatically

yuzefovich opened the pull request:
Note that it has been quite difficult to come up with a regression test,
yet I confirmed manually by augmenting the code to inject a retryable
error after the first run of the stats flow and observing the expected
error before the patch and the successful completion after the patch.

Fixes: #68670.

Release note (bug fix): Previously, table stats collection issued via
`ANALYZE` statement or via `CREATE STATISTICS` statement without
specifying AS OF SYSTEM TIME option could run into
`flow: memory budget exceeded`, and this has been fixed.

Release justification: fix to a long standing bug.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/69177)
<!-- Reviewable:end -->

yuzefovich commented:
@ajwerner I have a regression test set up that injects the retryable error as we wanted, but it seems to deadlock and I don’t understand why. Do you see anything immediately wrong?

Probably relevant goroutines:
```
goroutine 1214 [select]:
github.com/cockroachdb/cockroach/pkg/sql.(*createStatsNode).Next(0xc000bcd3b0, 0x933a4d0, 0xc001927290, 0xc002108e00, 0xc00105c5b0, 0x7b98520, 0x929c3e0, 0xc000c256d0)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/sql/create_stats.go:107 +0xbf
github.com/cockroachdb/cockroach/pkg/sql.(*planNodeToRowSource).Next(0xc0028fec00, 0x1, 0x0, 0xb4038e0, 0x0)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/sql/plan_node_to_row_source.go:173 +0x442
github.com/cockroachdb/cockroach/pkg/sql/colexec.(*Columnarizer).Next(0xc000a8f800, 0xc001ccfd00, 0x4098226)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/columnarizer.go:222 +0x10b
...
```
```
goroutine 1465 [select]:
github.com/cockroachdb/cockroach/pkg/sql.(*ieResultChannel).firstResult(0xc001eec100, 0x931ed58, 0xc0025cccf0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/sql/internal_result_channel.go:114 +0x114
github.com/cockroachdb/cockroach/pkg/sql.(*ieResultChannel).nextResult(0xc001eec100, 0x931ed58, 0xc0025cccf0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/sql/internal_result_channel.go:151 +0x14f
github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).execInternal(0xc0011c6600, 0x931ecb0, 0xc0025cccf0, 0x7f5bee0, 0xe, 0xc001eec100, 0xc000ba5ce0, 0x7f43f3d, 0x4, 0x0, ...)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/sql/internal.go:804 +0xcb2
github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).queryInternalBuffered(0xc0011c6600, 0x931ecb0, 0xc002261fc0, 0x7f5bee0, 0xe, 0xc000ba5ce0, 0x7f43f3d, 0x4, 0x0, 0x0, ...)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/sql/internal.go:425 +0x195
github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).QueryRowExWithCols(0xc0011c6600, 0x931ecb0, 0xc002261fc0, 0x7f5bee0, 0xe, 0xc000ba5ce0, 0x7f43f3d, 0x4, 0x0, 0x0, ...)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/sql/internal.go:483 +0x10b
github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).QueryRowEx(0xc0011c6600, 0x931ecb0, 0xc002261fc0, 0x7f5bee0, 0xe, 0xc000ba5ce0, 0x7f43f3d, 0x4, 0x0, 0x0, ...)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/sql/internal.go:469 +0xff
github.com/cockroachdb/cockroach/pkg/jobs.(*Job).load.func1(0x931ecb0, 0xc002261fc0, 0xc000ba5ce0, 0xc0021cc900, 0x3ff0000000000000)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/jobs/jobs.go:776 +0x1c4
github.com/cockroachdb/cockroach/pkg/kv.(*DB).Txn.func1(0x931ecb0, 0xc002261fc0, 0xc000ba5ce0, 0xc0021cc900, 0x3ff0000000000000)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/kv/db.go:839 +0x43
github.com/cockroachdb/cockroach/pkg/kv.(*Txn).exec(0xc000ba5ce0, 0x931ecb0, 0xc002261fc0, 0xc001914de8, 0xc000ba5ce0, 0x0)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/kv/txn.go:879 +0xdc
github.com/cockroachdb/cockroach/pkg/kv.(*DB).Txn(0xc0005b6f80, 0x931ecb0, 0xc002261fc0, 0xc002328e98, 0x140, 0x7cc5660)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/kv/db.go:838 +0x105
github.com/cockroachdb/cockroach/pkg/jobs.(*Job).runInTxn(0xc0002f2640, 0x931ecb0, 0xc002261fc0, 0x0, 0xc002328e98, 0xc0002f2640, 0x54320095b406357)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/jobs/jobs.go:751 +0x9e
github.com/cockroachdb/cockroach/pkg/jobs.(*Job).load(0xc0002f2640, 0x931ecb0, 0xc002261fc0, 0x0, 0xb4038e0, 0x1)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/jobs/jobs.go:774 +0xc5
github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).LoadJobWithTxn(0xc00000c780, 0x931ecb0, 0xc002261fc0, 0x9850cab0c7e8001, 0x0, 0x0, 0x0, 0x0)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/jobs/registry.go:692 +0x85
github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).LoadJob(...)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/jobs/registry.go:679
github.com/cockroachdb/cockroach/pkg/sql/rowexec.(*sampleAggregator).mainLoop(0xc0013f8c80, 0x931ecb0, 0xc002261fc0, 0xc00030b901, 0x7f65e81, 0x11)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/sql/rowexec/sample_aggregator.go:214 +0xd74
github.com/cockroachdb/cockroach/pkg/sql/rowexec.(*sampleAggregator).Run(0xc0013f8c80, 0x931ecb0, 0xc002261fc0)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/sql/rowexec/sample_aggregator.go:190 +0xd9
github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowBase).Run(0xc001070160, 0x931ecb0, 0xc002261fc0, 0x8597518)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow.go:418 +0x2c4
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run(0xc000d26b40, 0xc002126a80, 0xc001411a20, 0xc000e38000, 0xc000bf6700, 0xc002268240, 0x0, 0x0)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:525 +0x4d7
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).planAndRunCreateStats(0xc000d26b40, 0x931ed58, 0xc0025cc7b0, 0xc002268240, 0xc002126a80, 0xc001411a20, 0xc001d7bcc0, 0xc000f2b840, 0x0, 0x0)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_plan_stats.go:301 +0x2dd
github.com/cockroachdb/cockroach/pkg/sql.(*createStatsResumer).Resume.func1(0x931ed58, 0xc000bc8150, 0xc001411a20, 0x0, 0x0)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/sql/create_stats.go:563 +0x1a8
github.com/cockroachdb/cockroach/pkg/kv.(*DB).Txn.func1(0x931ed58, 0xc000bc8150, 0xc001411a20, 0x929d0e0, 0xc00030b540)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/kv/db.go:839 +0x43
github.com/cockroachdb/cockroach/pkg/kv.(*Txn).exec(0xc001411a20, 0x931ed58, 0xc000bc8150, 0xc001916d58, 0xc001411a20, 0xc0016d0f00)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/kv/txn.go:879 +0xdc
github.com/cockroachdb/cockroach/pkg/kv.(*DB).Txn(0xc0005b6f80, 0x931ed58, 0xc000bc8150, 0xc001b5edf0, 0xc000c25898, 0xc000ecbb18)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/kv/db.go:838 +0x105
github.com/cockroachdb/cockroach/pkg/sql.(*createStatsResumer).Resume(0xc001f691d0, 0x931ed58, 0xc000bc8150, 0x7cb8fc0, 0xc000c25898, 0x0, 0x0)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/sql/create_stats.go:539 +0x273
github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).stepThroughStateMachine.func1(0xc0005b9c00, 0x92e0c50, 0xc001f691d0, 0x931ed58, 0xc000bc8150, 0x7cb8fc0, 0xc000c25898, 0xc0019179e0)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/jobs/registry.go:1187 +0xae
github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).stepThroughStateMachine(0xc00000c780, 0x931ed58, 0xc000bc80f0, 0x7cb8fc0, 0xc000c25898, 0x92e0c50, 0xc001f691d0, 0xc001d7bcc0, 0x7f4a283, 0x7, ...)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/jobs/registry.go:1188 +0x8d2
github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).runJob(0xc00000c780, 0x931ecb0, 0xc001456cc0, 0x92e0c50, 0xc001f691d0, 0xc001d7bcc0, 0x7f4a283, 0x7, 0xc000ecb9e0, 0x16, ...)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/jobs/adopt.go:419 +0x46c
github.com/cockroachdb/cockroach/pkg/jobs.(*StartableJob).Start.func2(0x933a4d0, 0xc001f441b0)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/jobs/jobs.go:901 +0xf6
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2(0xc0005b6300, 0x933a4d0, 0xc001f441b0, 0xc001f44000, 0x0, 0xc001f69210)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:442 +0xf3
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:434 +0x22b
```
```
goroutine 1496 [select]:
github.com/cockroachdb/cockroach/pkg/kv/kvserver/txnwait.(*Queue).MaybeWaitForPush(0xc000d2b310, 0x931ed58, 0xc00253bfb0, 0xc0015a8c40, 0x0, 0x0)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/txnwait/queue.go:513 +0x10dc
github.com/cockroachdb/cockroach/pkg/kv/kvserver/concurrency.(*managerImpl).maybeInterceptReq(0xc000d2b270, 0x931ed58, 0xc00253bfb0, 0x0, 0x169cd4fefb50fb18, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/concurrency/concurrency_manager.go:330 +0xa6
github.com/cockroachdb/cockroach/pkg/kv/kvserver/concurrency.(*managerImpl).sequenceReqWithGuard(0xc000d2b270, 0x931ed58, 0xc00253bfb0, 0xc001c4a5a0, 0x411c21c, 0xc0025cd770, 0x0, 0x0)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/concurrency/concurrency_manager.go:236 +0x138
github.com/cockroachdb/cockroach/pkg/kv/kvserver/concurrency.(*managerImpl).SequenceReq(0xc000d2b270, 0x931ed58, 0xc00253bfb0, 0xc001c4a5a0, 0x0, 0x169cd4fefb50fb18, 0x0, 0x0, 0x0, 0x0, ...)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/concurrency/concurrency_manager.go:210 +0xe5
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).executeBatchWithConcurrencyRetries(0xc001bc8a80, 0x931ed58, 0xc00253bfb0, 0xc00107d0a0, 0x8596728, 0x0, 0x0)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_send.go:357 +0x2e8
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).sendWithRangeID(0xc001bc8a80, 0x931ed58, 0xc00253bf80, 0xb, 0xc00107d0a0, 0x0, 0x0)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_send.go:107 +0x59d
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).Send(...)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_send.go:43
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Store).Send(0xc001762800, 0x931ed58, 0xc00253bf50, 0x169cd4fefb50fb18, 0x0, 0x0, 0x100000001, 0x1, 0x0, 0xb, ...)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/store_send.go:190 +0x62c
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Stores).Send(0xc000163c00, 0x931ed58, 0xc00253bf50, 0x169cd4fefb50fb18, 0x0, 0x0, 0x100000001, 0x1, 0x0, 0xb, ...)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/stores.go:191 +0x11e
github.com/cockroachdb/cockroach/pkg/server.(*Node).batchInternal.func1(0x931ed58, 0xc00253bf50, 0x0, 0x0)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/server/node.go:906 +0x225
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunTaskWithErr(0xc0005b6300, 0x931ed58, 0xc00253bf50, 0x7f6235a, 0x10, 0xc0022b9be8, 0x0, 0x0)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:324 +0xb2
github.com/cockroachdb/cockroach/pkg/server.(*Node).batchInternal(0xc000261800, 0x931ed58, 0xc00253bf50, 0xc00107cfc0, 0x0, 0x0, 0x1)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/server/node.go:894 +0x165
github.com/cockroachdb/cockroach/pkg/server.(*Node).Batch(0xc000261800, 0x931ed58, 0xc00253bf20, 0xc00107cfc0, 0xc0025cd800, 0x0, 0x0)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/server/node.go:987 +0xd2
github.com/cockroachdb/cockroach/pkg/rpc.internalClientAdapter.Batch(...)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/rpc/context.go:469
github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*grpcTransport).sendBatch(0xc0003aadc0, 0x931ed58, 0xc00253bf20, 0x1, 0x936d618, 0xc00041d570, 0x169cd4fefb50fb18, 0x0, 0x0, 0x100000001, ...)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/transport.go:195 +0x176
github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*grpcTransport).SendNext(0xc0003aadc0, 0x931ed58, 0xc0025cd800, 0x169cd4fefb50fb18, 0x0, 0x0, 0x100000001, 0x1, 0x0, 0xb, ...)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/transport.go:177 +0x198
github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*DistSender).sendToReplicas(0xc000b80480, 0x931ed58, 0xc0025cd800, 0x169cd4fefb50fb18, 0x0, 0x0, 0x0, 0x0, 0x0, 0xb, ...)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/dist_sender.go:1931 +0xaf7
github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*DistSender).sendPartialBatch(0xc000b80480, 0x931ed58, 0xc0025cd800, 0x169cd4fefb50fb18, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/dist_sender.go:1526 +0x35f
github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*DistSender).divideAndSendBatchToRanges(0xc000b80480, 0x931ed58, 0xc0025cd800, 0x169cd4fefb50fb18, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/dist_sender.go:1144 +0x388
github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*DistSender).Send(0xc000b80480, 0x931ed58, 0xc0025cd800, 0x169cd4fefb50fb18, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/dist_sender.go:778 +0x61b
github.com/cockroachdb/cockroach/pkg/kv.(*CrossRangeTxnWrapperSender).Send(0xc0005b6fd8, 0x931ed58, 0xc0025cd800, 0x169cd4fefb50fb18, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/kv/db.go:220 +0xb5
github.com/cockroachdb/cockroach/pkg/kv.(*DB).sendUsingSender(0xc0005b6f80, 0x931ed58, 0xc0025cd800, 0x169cd4fefb50fb18, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/kv/db.go:875 +0x142
github.com/cockroachdb/cockroach/pkg/kv.(*DB).send(...)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/kv/db.go:858
github.com/cockroachdb/cockroach/pkg/kv.sendAndFill(0x931ed58, 0xc0025cd800, 0xc0022bc520, 0xc002083400, 0x0, 0xc0022bca78)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/kv/db.go:786 +0x155
github.com/cockroachdb/cockroach/pkg/kv.(*DB).Run(0xc0005b6f80, 0x931ed58, 0xc0025cd800, 0xc002083400, 0xc0022bc7a0, 0x1)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/kv/db.go:809 +0xa5
github.com/cockroachdb/cockroach/pkg/kv/kvserver/intentresolver.(*IntentResolver).MaybePushTransactions(0xc0001640e0, 0x931ed58, 0xc0025cd800, 0xc0022bca78, 0x169cd4fefb50c468, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/intentresolver/intent_resolver.go:389 +0x778
github.com/cockroachdb/cockroach/pkg/kv/kvserver/intentresolver.(*IntentResolver).PushTransaction(0xc0001640e0, 0x931ed58, 0xc0025cd800, 0xc0002bd0b8, 0x169cd4fefb50c468, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/intentresolver/intent_resolver.go:297 +0x14e
github.com/cockroachdb/cockroach/pkg/kv/kvserver/concurrency.(*lockTableWaiterImpl).pushLockTxn(0xc000d2b2c0, 0x931ed58, 0xc0025cd800, 0xc0021ccea0, 0x169cd4fef84f24f8, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/concurrency/lock_table_waiter.go:519 +0x13c
github.com/cockroachdb/cockroach/pkg/kv/kvserver/concurrency.(*lockTableWaiterImpl).WaitOn.func3(0x931ed58, 0xc0025cd800, 0x0)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/concurrency/lock_table_waiter.go:372 +0x165
github.com/cockroachdb/cockroach/pkg/kv/kvserver/concurrency.(*lockTableWaiterImpl).WaitOn(0xc000d2b2c0, 0x931ed58, 0xc0025cd800, 0xc0021ccea0, 0x169cd4fef84f24f8, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/concurrency/lock_table_waiter.go:423 +0x510
github.com/cockroachdb/cockroach/pkg/kv/kvserver/concurrency.(*managerImpl).sequenceReqWithGuard(0xc000d2b270, 0x931ed58, 0xc0025cd800, 0xc00236e320, 0x16, 0x797f380, 0x0, 0x1)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/concurrency/concurrency_manager.go:311 +0x482
github.com/cockroachdb/cockroach/pkg/kv/kvserver/concurrency.(*managerImpl).SequenceReq(0xc000d2b270, 0x931ed58, 0xc0025cd800, 0xc00236e320, 0xc0021ccea0, 0x169cd4fef84f24f8, 0x0, 0x0, 0x0, 0x0, ...)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/concurrency/concurrency_manager.go:210 +0xe5
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).executeBatchWithConcurrencyRetries(0xc001bc8a80, 0x931ed58, 0xc0025cd800, 0xc002126d20, 0x8596710, 0x0, 0x0)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_send.go:357 +0x2e8
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).sendWithRangeID(0xc001bc8a80, 0x931ed58, 0xc0025cd7d0, 0xb, 0xc002126d20, 0x0, 0x1)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_send.go:103 +0x2d2
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).Send(...)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_send.go:43
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Store).Send(0xc001762800, 0x931ed58, 0xc0025cd7a0, 0x169cd4fef84f24f8, 0x0, 0x0, 0x100000001, 0x1, 0x0, 0xb, ...)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/store_send.go:190 +0x62c
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Stores).Send(0xc000163c00, 0x931ed58, 0xc0025cd7a0, 0x0, 0x0, 0x0, 0x100000001, 0x1, 0x0, 0xb, ...)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/stores.go:191 +0x11e
github.com/cockroachdb/cockroach/pkg/server.(*Node).batchInternal.func1(0x931ed58, 0xc0025cd7a0, 0x0, 0x0)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/server/node.go:906 +0x225
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunTaskWithErr(0xc0005b6300, 0x931ed58, 0xc0025cd7a0, 0x7f6235a, 0x10, 0xc0022a91f0, 0x0, 0x0)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:324 +0xb2
github.com/cockroachdb/cockroach/pkg/server.(*Node).batchInternal(0xc000261800, 0x931ed58, 0xc0025cd7a0, 0xc002126c40, 0x0, 0x0, 0x1)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/server/node.go:894 +0x165
github.com/cockroachdb/cockroach/pkg/server.(*Node).Batch(0xc000261800, 0x931ed58, 0xc0025cd770, 0xc002126c40, 0xc001eec500, 0x0, 0x0)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/server/node.go:987 +0xd2
github.com/cockroachdb/cockroach/pkg/rpc.internalClientAdapter.Batch(...)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/rpc/context.go:469
github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*grpcTransport).sendBatch(0xc00260e000, 0x931ed58, 0xc0025cd770, 0x1, 0x936d618, 0xc00041d570, 0x0, 0x0, 0x0, 0x100000001, ...)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/transport.go:195 +0x176
github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*grpcTransport).SendNext(0xc00260e000, 0x931ecb0, 0xc001eec500, 0x0, 0x0, 0x0, 0x100000001, 0x1, 0x0, 0xb, ...)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/transport.go:177 +0x198
github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*DistSender).sendToReplicas(0xc000b80480, 0x931ecb0, 0xc001eec500, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0xb, ...)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/dist_sender.go:1931 +0xaf7
github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*DistSender).sendPartialBatch(0xc000b80480, 0x931ecb0, 0xc001eec500, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/dist_sender.go:1526 +0x35f
github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*DistSender).divideAndSendBatchToRanges(0xc000b80480, 0x931ecb0, 0xc001eec500, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/dist_sender.go:1144 +0x388
github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*DistSender).Send(0xc000b80480, 0x931ecb0, 0xc001eec500, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/dist_sender.go:778 +0x61b
github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*txnLockGatekeeper).SendLocked(0xc002195358, 0x931ecb0, 0xc001eec500, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/txn_lock_gatekeeper.go:82 +0x175
github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*txnMetricRecorder).SendLocked(0xc002195320, 0x931ecb0, 0xc001eec500, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/txn_interceptor_metric_recorder.go:46 +0x9e
github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*txnCommitter).SendLocked(0xc0021952f0, 0x931ecb0, 0xc001eec500, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/txn_interceptor_committer.go:127 +0x995
github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*txnSpanRefresher).sendLockedWithRefreshAttempts(0xc002195250, 0x931ecb0, 0xc001eec500, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/txn_interceptor_span_refresher.go:240 +0xb5
github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*txnSpanRefresher).SendLocked(0xc002195250, 0x931ecb0, 0xc001eec500, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/txn_interceptor_span_refresher.go:175 +0x285
github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*txnPipeliner).SendLocked(0xc002195180, 0x931ecb0, 0xc001eec500, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/txn_interceptor_pipeliner.go:285 +0x335
github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*txnSeqNumAllocator).SendLocked(0xc002195160, 0x931ecb0, 0xc001eec500, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/txn_interceptor_seq_num_allocator.go:105 +0x215
github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*txnHeartbeater).SendLocked(0xc0021950c0, 0x931ecb0, 0xc001eec500, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/txn_interceptor_heartbeater.go:240 +0x295
github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*TxnCoordSender).Send(0xc002194f00, 0x931ecb0, 0xc001eec500, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/txn_coord_sender.go:513 +0x302
github.com/cockroachdb/cockroach/pkg/kv.(*DB).sendUsingSender(0xc0005b6f80, 0x931ecb0, 0xc001eec500, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/kv/db.go:875 +0x142
github.com/cockroachdb/cockroach/pkg/kv.(*Txn).Send(0xc000ba5ce0, 0x931ecb0, 0xc001eec500, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/kv/txn.go:993 +0x168
github.com/cockroachdb/cockroach/pkg/sql/row.makeKVBatchFetcherDefaultSendFunc.func1(0x931ecb0, 0xc001eec500, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/sql/row/kv_batch_fetcher.go:224 +0x98
github.com/cockroachdb/cockroach/pkg/sql/row.(*txnKVFetcher).fetch(0xc0011642d0, 0x931ecb0, 0xc001eec500, 0x633365e, 0xc0025cd680)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/sql/row/kv_batch_fetcher.go:442 +0x935
github.com/cockroachdb/cockroach/pkg/sql/row.(*txnKVFetcher).nextBatch(0xc0011642d0, 0x931ecb0, 0xc001eec500, 0x50, 0xb, 0x94478a0, 0x6e059f4, 0xc0021913b8, 0x0, 0x0, ...)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/sql/row/kv_batch_fetcher.go:588 +0x749
github.com/cockroachdb/cockroach/pkg/sql/row.(*KVFetcher).NextKV(0xc00028ed20, 0x931ecb0, 0xc001eec500, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/sql/row/kv_fetcher.go:176 +0xd9
github.com/cockroachdb/cockroach/pkg/sql/colfetcher.(*cFetcher).NextBatch(0xc001eeb440, 0x931ecb0, 0xc001eec500, 0x0, 0x0, 0x0, 0x0)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/sql/colfetcher/cfetcher.go:848 +0xb9
github.com/cockroachdb/cockroach/pkg/sql/colfetcher.(*ColBatchScan).Next(0xc001d79710, 0x0, 0x0)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/sql/colfetcher/colbatch_scan.go:110 +0x45
github.com/cockroachdb/cockroach/pkg/sql/colexec.(*InvariantsChecker).Next(0xc001eec600, 0x0, 0xc000db35e0)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/invariants_checker.go:85 +0x69
github.com/cockroachdb/cockroach/pkg/sql/colexec/colexecutils.(*CancelChecker).Next(0xc001eec640, 0xc001f76060, 0x1)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/colexecutils/cancel_checker.go:59 +0x42
github.com/cockroachdb/cockroach/pkg/sql/colexec/colexecbase.(*simpleProjectOp).Next(0xc001213da0, 0x0, 0xc0005b8301)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/colexecbase/simple_project.go:125 +0x49
github.com/cockroachdb/cockroach/pkg/sql/colexec.(*InvariantsChecker).Next(0xc001eec680, 0x7954301, 0xc0022adc20)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/invariants_checker.go:85 +0x69
github.com/cockroachdb/cockroach/pkg/sql/colexec.(*Materializer).next(0xc001070b00, 0xc001eec500, 0xc0025cd301, 0xc0022adc48)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/materializer.go:261 +0x14b
github.com/cockroachdb/cockroach/pkg/sql/colexec.(*Materializer).nextAdapter(...)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/materializer.go:286
github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError(0xc0022adc88, 0x0, 0x0)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexecerror/error.go:91 +0x57
github.com/cockroachdb/cockroach/pkg/sql/colexec.(*Materializer).Next(0xc001070b00, 0x0, 0x931ecb0, 0xc001eec500, 0xc001eec500)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/materializer.go:292 +0x58
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*FlowCoordinator).next(0xc00170fd00, 0xc001eec500, 0xc0022add20, 0x6e06fcb, 0xc001070b00)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/sql/colflow/flow_coordinator.go:127 +0x73
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*FlowCoordinator).nextAdapter(...)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/sql/colflow/flow_coordinator.go:144
github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError(0xc0022add70, 0x0, 0x0)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexecerror/error.go:91 +0x57
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*FlowCoordinator).Next(0xc00170fd00, 0xc001eec500, 0x6e06f80, 0xc00170fd00, 0x931ecb0)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/sql/colflow/flow_coordinator.go:149 +0x4d
github.com/cockroachdb/cockroach/pkg/sql/execinfra.Run(0x931ecb0, 0xc001eec500, 0x933e7f8, 0xc00170fd00, 0x92e0c00, 0xc0013e5500)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfra/base.go:186 +0x35
github.com/cockroachdb/cockroach/pkg/sql/execinfra.(*ProcessorBaseNoHelper).Run(0xc00170fd00, 0x931ecb0, 0xc001eec500)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfra/processorsbase.go:731 +0x90
github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowBase).Run(0xc0010709a0, 0x931ecb0, 0xc001eec500, 0x8597518)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow.go:418 +0x2c4
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*vectorizedFlow).Run(0xc001929b30, 0x931ecb0, 0xc001eec500, 0x8597518)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/sql/colflow/vectorized_flow.go:244 +0x28a
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run(0xc000d26b40, 0xc002126b60, 0xc000ba5ce0, 0xc000e38580, 0xc0013e5500, 0xc00219a7f0, 0x0, 0x0)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:525 +0x4d7
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRun(0xc000d26b40, 0x931ed58, 0xc0025cd3e0, 0xc00219a7f0, 0xc002126b60, 0xc000ba5ce0, 0x93208e8, 0xc001d238c0, 0x0, 0xc0013e5500, ...)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:1262 +0x1cb
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithDistSQLEngine(0xc00219a000, 0x931ed58, 0xc0025cd3e0, 0xc00219a5b0, 0x3, 0x9446b10, 0xc0025cd3b0, 0xc002192800, 0xc001213978, 0x0, ...)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1153 +0x468
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).dispatchToExecutionEngine(0xc00219a000, 0x931ed58, 0xc0025cd3e0, 0xc00219a5b0, 0x9446b10, 0xc0025cd3b0, 0x0, 0x0)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1015 +0x8e6
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState(0xc00219a000, 0x931ed58, 0xc0025cd3e0, 0x93450b8, 0xc000dda460, 0x8066d16, 0x57, 0x1, 0x1, 0xc000859ef0, ...)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:706 +0x1046
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt(0xc00219a000, 0x931ecb0, 0xc001eec2c0, 0x93450b8, 0xc000dda460, 0x8066d16, 0x57, 0x1, 0x1, 0xc000859ef0, ...)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:127 +0xb8f
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execPortal(0xc00219a000, 0x931ecb0, 0xc001eec2c0, 0xc000859ef0, 0xc0021e05d0, 0x1, 0x1, 0x0, 0x0, 0x0, ...)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:214 +0x14e
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd.func2(0x0, 0x0, 0x0, 0x2cf12080, 0xed8b0cf1c, 0x0, 0xc00219a000, 0xc0029bfbc0, 0xc0029bfbb0, 0xc002193ab8, ...)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1708 +0x4a9
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd(0xc00219a000, 0x931ecb0, 0xc001eec2c0, 0x0, 0x0)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1710 +0x4df
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run(0xc00219a000, 0x931ed58, 0xc0025cccf0, 0xc00017bd60, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1564 +0x225
github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).initConnEx.func1(0xc00219a000, 0x931ed58, 0xc0025cccf0, 0xc0011c6600, 0xc0025cce10, 0x92e0ca0, 0xc001eec100, 0xc000ba5ce0, 0xc001f61480)
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/sql/internal.go:204 +0x65
created by github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).initConnEx
	/Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/sql/internal.go:203 +0x305
```
The last one seems to be the culprit. I wonder whether I somehow messed up the timestamps or something like that.

michae2 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69177)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @yuzefovich)

___
*[pkg/sql/create_stats.go, line 557 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/69177#-Mht90465WyH-3rH1dGH:-Mht90465WyH-3rH1dGI:b-qi0y09) ([raw file](https://github.com/cockroachdb/cockroach/blob/7d143e202c835748107bdf44c2bbfceff90d2f2f/pkg/sql/create_stats.go#L557)):*
> ```
> 			p.ExtendedEvalContext().AsOfSystemTime = &tree.AsOfSystemTime{Timestamp: *details.AsOf}
> 			p.ExtendedEvalContext().SetTxnTimestamp(details.AsOf.GoTime())
> ```

Do these need to be changed to act on `evalCtx` instead of `p.ExtendedEvalContext()`?


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69177)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @michae2)

___
*[pkg/sql/create_stats.go, line 557 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/69177#-Mht90465WyH-3rH1dGH:-MhuvGkN9bPipQBrSrk5:b6wmbcf) ([raw file](https://github.com/cockroachdb/cockroach/blob/7d143e202c835748107bdf44c2bbfceff90d2f2f/pkg/sql/create_stats.go#L557)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

> ```
> 			p.ExtendedEvalContext().AsOfSystemTime = &tree.AsOfSystemTime{Timestamp: *details.AsOf}
> 			p.ExtendedEvalContext().SetTxnTimestamp(details.AsOf.GoTime())
> ```

Do these need to be changed to act on `evalCtx` instead of `p.ExtendedEvalContext()`?
</blockquote></details>

Yes, very good point, fixed.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
@cockroachdb/sql-queries thoughts on merging the fix without a regression test (i.e. only the currently second commit)? I don't understand why my attempt at the regression test deadlocks, and Drew has also spent noticeable amount of time trying to come up a with a regression test. The bug seems clear enough that I'm thinking that it might be not worth scratching our head for much longer for the regression test.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69177#-:-MhuvLQt6YXvT1mA_GJa:b-80l3s3)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @michae2)



<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69177)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @michae2 and @yuzefovich)

___
*[pkg/sql/create_stats.go, line 550 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/69177#-Mhv3U-53BoairUFvVAV:-Mhv3U-6389UVtrNVa4R:b-ntkn4f) ([raw file](https://github.com/cockroachdb/cockroach/blob/289969f07401b96b998eea3721ecc4e345d04aad/pkg/sql/create_stats.go#L550)):*
> ```Go
> 		// Make a copy of the EvalContext because the flow setup mutates it
> 		// (namely, it updates .Mon field).
> 		evalCtx.EvalContext = *p.ExtendedEvalContext().EvalContext.Copy()
> ```

Really confused what we are doing here. `evalCtx` is an `*extendedEvalContext` which embeds `EvalContext`. How does this change help with some later code that is modifying `evalCtx.Mon`?

I see there is a `ExtendedEvalContextCopy()`, maybe we should just use that instead of `ExtendedEvalContext` above?


<!-- Sent from Reviewable.io -->

rytaft reviewed (commented):
Is there a way to confirm manually that this fixes the problem? e.g., run ANALYZE/CREATE STATISTICS a bunch of times in a row on a big table or set of tables?

> Reviewed 1 of 3 files at r1, 2 of 2 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69177#-:-Mhxu5bX87Lvxu7pS7dV:b-6hluzz)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @michae2 and @yuzefovich)

___
*[-- commits, line 7 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/69177#-MhxqyO20gdRzpIat_kN):*
Isn't `ANALYZE` already fixed due to @mgartner's change?
___
*[pkg/sql/stats/create_stats_job_test.go, line 539 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/69177#-MhxqhMJ4KjVC0vNbuPV:-MhxqhMK7E6Fc2Cpso1q:b-u50jmz) ([raw file](https://github.com/cockroachdb/cockroach/blob/f89909c67848a6facbf6ebcdd39f23db297574bd/pkg/sql/stats/create_stats_job_test.go#L539)):*
> ```Go
> 	sqlDB.Exec(t, `CREATE TABLE t (x INT PRIMARY KEY)`)
> 	sqlDB.Exec(t, `INSERT INTO t SELECT generate_series(1,1000)`)
> 	sqlDB.Exec(t, `ANALYZE t`)
> ```

Now that @mgartner changed `ANALYZE` to use the inconsistent scan, maybe you should check `CREATE STATISTICS` with no AOST?


<!-- Sent from Reviewable.io -->

yuzefovich committed:
sql: fix stats jobs when they are retried automatically

Note that it has been quite difficult to come up with a regression test,
yet I confirmed manually by augmenting the code to inject a retryable
error after the first run of the stats flow and observing the expected
error before the patch and the successful completion after the patch.

Release note (bug fix): Previously, table stats collection issued via
`ANALYZE` statement or via `CREATE STATISTICS` statement without
specifying AS OF SYSTEM TIME option could run into
`flow: memory budget exceeded`, and this has been fixed.

Release justification: fix to a long standing bug.

yuzefovich reviewed (commented):
I used the following manual test. I adjusted the binary with the following patch:
```
diff --git a/pkg/sql/create_stats.go b/pkg/sql/create_stats.go
index 095ab520b8..23e02740ef 100644
--- a/pkg/sql/create_stats.go
+++ b/pkg/sql/create_stats.go
@@ -18,6 +18,7 @@ import (
        "github.com/cockroachdb/cockroach/pkg/jobs"
        "github.com/cockroachdb/cockroach/pkg/jobs/jobspb"
        "github.com/cockroachdb/cockroach/pkg/kv"
+       "github.com/cockroachdb/cockroach/pkg/roachpb"
        "github.com/cockroachdb/cockroach/pkg/server/telemetry"
        "github.com/cockroachdb/cockroach/pkg/settings"
        "github.com/cockroachdb/cockroach/pkg/settings/cluster"
@@ -535,7 +536,24 @@ func (r *createStatsResumer) Resume(ctx context.Context, execCtx interface{}) er
        evalCtx := p.ExtendedEvalContext()
 
        dsp := p.DistSQLPlanner()
-       if err := p.ExecCfg().DB.Txn(ctx, func(ctx context.Context, txn *kv.Txn) error {
+       alreadyInjected := false
+       if err := p.ExecCfg().DB.Txn(ctx, func(ctx context.Context, txn *kv.Txn) (retErr error) {
+               defer func() {
+                       if retErr == nil && !alreadyInjected {
+                               clock := p.ExecCfg().Clock
+                               retErr = roachpb.NewTransactionRetryWithProtoRefreshError(
+                                       "after injected retryable error",
+                                       txn.ID(),
+                                       roachpb.MakeTransaction(
+                                               txn.DebugName(),
+                                               nil, // baseKey
+                                               txn.UserPriority(),
+                                               clock.NowAsClockTimestamp().ToTimestamp(),
+                                               clock.MaxOffset().Nanoseconds(),
+                                       ))
+                               alreadyInjected = true
+                       }
+               }()
                // Make the copy of the extended eval context since we're modifying it
                // below. Notably, evalCtxCopy.EvalContext.Mon is overwritten during the
                // flow setup, and we want to use the original monitor in case DB.Txn is
```
to inject a retryable error after the stats collection flow completes for the first time. Before the patch I got the "flow budget exceeded" error when analyzing `lineitem` table, and after the patch the analyze succeeded.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69177#-:-MhzKxow2gcej88djeHU:b-sbpofi)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mgartner, @michae2, @RaduBerinde, and @rytaft)

___
*[-- commits, line 7 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/69177#-MhxqyO20gdRzpIat_kN):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Isn't `ANALYZE` already fixed due to @mgartner's change?
</blockquote></details>

Hm, I'm not sure. Using AOST definitely reduces the likelihood of encountering a retryable txn error, but I'm not sure whether the usage of inconsistent scans makes it so that a retryable txn error never occurs.
___
*[pkg/sql/create_stats.go, line 550 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/69177#-Mhv3U-53BoairUFvVAV:-MhzKlDaAfgJNPUXGOwv:b-48x666) ([raw file](https://github.com/cockroachdb/cockroach/blob/289969f07401b96b998eea3721ecc4e345d04aad/pkg/sql/create_stats.go#L550)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Really confused what we are doing here. `evalCtx` is an `*extendedEvalContext` which embeds `EvalContext`. How does this change help with some later code that is modifying `evalCtx.Mon`?

I see there is a `ExtendedEvalContextCopy()`, maybe we should just use that instead of `ExtendedEvalContext` above?
</blockquote></details>

I think you're right that my original "fix" didn't work, and your suggestion works and is simpler.
___
*[pkg/sql/stats/create_stats_job_test.go, line 539 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/69177#-MhxqhMJ4KjVC0vNbuPV:-MhzKfpi2BLb-rhNN8J3:b-s9n6ch) ([raw file](https://github.com/cockroachdb/cockroach/blob/f89909c67848a6facbf6ebcdd39f23db297574bd/pkg/sql/stats/create_stats_job_test.go#L539)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Now that @mgartner changed `ANALYZE` to use the inconsistent scan, maybe you should check `CREATE STATISTICS` with no AOST?
</blockquote></details>

I removed the test since I don't know how to make it not deadlock.


<!-- Sent from Reviewable.io -->

rytaft reviewed (approved):
Ok, that sounds like a good test. Can you also add that this fixes #67113?

I think it's probably fine to go ahead with this since it fixes a known bug, but it seems like maybe there's a bigger problem here? Why are we overwriting the Mon during flow setup? Wouldn't that cause problems for other transactions with retryable errors too? And is there any downside to copying the eval context like you're doing here?

> Reviewed 3 of 3 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69177#-:-Mi6fNvO6sA2-a79XiJh:bdyafhk)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @michae2 and @RaduBerinde)



<!-- Sent from Reviewable.io -->
